server/src/main/java/io/crate/execution/engine/aggregation/impl/TopKAggregation.java

```diff
@@ -0,0 +1,273 @@
+
+    public static final String NAME = "topk";
+
+    static final Signature DEFAULT_SIGNATURE =
+        Signature.builder(NAME, FunctionType.AGGREGATE)
+            .argumentTypes(TypeSignature.parse("V"))
+            .returnType(new ArrayType<>(DataTypes.UNTYPED_OBJECT).getTypeSignature())
+            .features(Scalar.Feature.DETERMINISTIC)
+            .typeVariableConstraints(typeVariable("V"))
+            .build();
+
+    static final Signature PARAMETER_SIGNATURE =
+        Signature.builder(NAME, FunctionType.AGGREGATE)
+            .argumentTypes(TypeSignature.parse("V"),
+                DataTypes.INTEGER.getTypeSignature())
+            .returnType(new ArrayType<>(DataTypes.UNTYPED_OBJECT).getTypeSignature())
+            .features(Scalar.Feature.DETERMINISTIC)
+            .typeVariableConstraints(typeVariable("V"))
+            .build();
+
+    static {
+        DataTypes.register(StateType.ID, in -> StateType.INSTANCE);
+    }
+
+    public static void register(Functions.Builder builder) {
+        builder.add(
+            DEFAULT_SIGNATURE,
+            TopKAggregation::new
+        );
+
+        builder.add(
+            PARAMETER_SIGNATURE,
+            TopKAggregation::new
+        );
+    }
+
+    private final Signature signature;
+    private final BoundSignature boundSignature;
+    private static final int DEFAULT_MAP_SIZE = 32;
+    private static final int DEFAULT_LIMIT = 8;
+    private static final int MAX_LIMIT = 10_000;
+
+    private TopKAggregation(Signature signature, BoundSignature boundSignature) {
+        this.signature = signature;
+        this.boundSignature = boundSignature;
+    }
+
+    @Override
+    public Signature signature() {
+        return signature;
+    }
+
+    @Override
+    public BoundSignature boundSignature() {
+        return boundSignature;
+    }
+
+
+    @Nullable
+    @Override
+    public State newState(RamAccounting ramAccounting,
+                          Version indexVersionCreated,
+                          Version minNodeInCluster,
+                          MemoryManager memoryManager) {
+        return State.EMPTY;
+    }
+
+    @Override
+    public State iterate(RamAccounting ramAccounting,
+                         MemoryManager memoryManager,
+                         State state,
+                         Input<?>... args) throws CircuitBreakingException {
+        Object value = args[0].value();
+        if (state instanceof Empty) {
+            if (args.length == 2) {
+                // We have a limit provided by the user
+                Integer limit = (Integer) args[1].value();
+                if (limit <= 0 || limit > MAX_LIMIT) {
+                    throw new IllegalArgumentException("Invalid value for parameter k for top-k aggregation");
+                }
+                int maxMapSize = maxMapSize(limit);
+                ramAccounting.addBytes(calculateRamUsage(maxMapSize));
+                DataType<?> dataType = DataTypes.guessType(value);
```

Comment:
The argument types are part of the bound signature. No need to do type guessing here.